Update docs to OS contributors (#685)
Stand-With-Crypto/swc-web: docs/Contributing.md

# Contributing Guide

## Questions

> If you have questions about implementation details, help or support, then please use our dedicated community forum at [GitHub Discussions](https://github.com/Stand-With-Crypto/swc-web/discussions) PLEASE NOTE: If you choose to instead open an issue for your question, your issue will be immediately closed and redirected to the forum.

Reviewer: Oxford comma - `implementation details, help, or support`

Reviewer: ```suggestion
If you have questions about implementation details, help, or support, then please use our dedicated community forum at [GitHub Discussions](https://github.com/Stand-With-Crypto/swc-web/discussions). PLEASE NOTE: If you choose to instead open an issue for your question, your issue will be immediately closed and redirected to the forum.
```

## Reporting Issues

> If you have found what you think is a bug, please [file an issue](https://github.com/Stand-With-Crypto/swc-web/issues/new). PLEASE NOTE: Issues that are identified as implementation questions or non-issues will be immediately closed and redirected to [GitHub Discussions](https://github.com/Stand-With-Crypto/swc-web/discussions)

Reviewer: If you have found what you think is a bug, please [file an issue](https://github.com/Stand-With-Crypto/swc-web/issues/new). PLEASE NOTE: Issues that are identified as implementation questions or non-issues will be immediately closed and redirected to [GitHub Discussions](https://github.com/Stand-With-Crypto/swc-web/discussions)
```suggestion
If you have found what you think is a bug, please [file an issue](https://github.com/Stand-With-Crypto/swc-web/issues/new). PLEASE NOTE: Issues that are identified as implementation questions or non-issues will be immediately closed and redirected to [GitHub Discussions](https://github.com/Stand-With-Crypto/swc-web/discussions).
```

### Required tools

Make sure you have the following tools installed before proceeding:

- [nvm](https://github.com/nvm-sh/nvm)
- [docker](https://docs.docker.com/compose/install/)
- [docker-compose](https://docs.docker.com/compose/install/#scenario-two-install-the-compose-plugin)

### Developing your feature/fix

If you have been assigned to fix an issue or develop a new feature, please follow these steps to get started:

- Fork this repository
  - If you are a core contributors clone it instead
- Install dependencies by running `$ npm install`
  - We use npm for managing dependencies
  - We use [nvm](https://github.com/nvm-sh/nvm) to manage node versions - please use the version mentioned in `.nvmrc` (run `$ nvm use`)
- Create a `.env` file based on `.env.example` (run `$ cp .env.example .env`)
  - If you are a core contributor you can install [Vercel CLI](https://vercel.com/docs/cli) and run `$ vercel link` and then `$ vercel env pull .env`
- Open the `.env` file and fill all the blank values that are not bellow the `OPTIONAL` title. Follow the commented instructions for each one
- Start the local db by running `$ docker-compose -f dev.docker-compose.yml up -d`

>   - We use docker-compose to run services both in local dev and in e2e testing

Reviewer: The Docker file for the local service and the Docker file for Cypress uses the same port for their MySQL database, right?

I think we should have a separate port for the Cypress MySQL database because some of the Cypress tests involve wiping the MySQL database to have a fresh database, and we probably do not want to unintentionally wipe the user's service database if they run the Cypress suite.

# Contributing Guide

## Questions

If you have questions about implementation details, help or support, then please use our dedicated community forum at [GitHub Discussions](https://github.com/Stand-With-Crypto/swc-web/discussions) PLEASE NOTE: If you choose to instead open an issue for your question, your issue will be immediately closed and redirected to the forum.

## Reporting Issues

If you have found what you think is a bug, please [file an issue](https://github.com/Stand-With-Crypto/swc-web/issues/new). PLEASE NOTE: Issues that are identified as implementation questions or non-issues will be immediately closed and redirected to [GitHub Discussions](https://github.com/Stand-With-Crypto/swc-web/discussions)

## Suggesting new features

If you are here to suggest a feature, first [create an issue](https://github.com/Stand-With-Crypto/swc-web/issues/new) if it does not already exist. From there, we will discuss use-cases for the feature and then finally discuss how it could be implemented.

## Development

### Knowledge pre-requisites

> Before proceeding with anything, **PLEASE** check out this following resources.

Reviewer 2: This was really only needed when I was onboarding new team members. we can soften the language:
```suggestion
Below is a list of suggested documentation to read through if you aren't already familiar with the technologies used in this repository:
```

- Read the [Inngest docs](https://www.inngest.com/docs/quick-start) to learn more about how to leverage the tool to build resilient workflows

### Required tools

Make sure you have the following tools installed before proceeding:

- [nvm](https://github.com/nvm-sh/nvm)
- [docker](https://docs.docker.com/compose/install/)
- [docker-compose](https://docs.docker.com/compose/install/#scenario-two-install-the-compose-plugin)

### Developing your feature/fix

If you have been assigned to fix an issue or develop a new feature, please follow these steps to get started:

- Fork this repository
  - If you are a core contributors clone it instead
- Install dependencies by running `$ npm install`
  - We use npm for managing dependencies
  - We use [nvm](https://github.com/nvm-sh/nvm) to manage node versions - please use the version mentioned in `.nvmrc` (run `$ nvm use`)
- Create a `.env` file based on `.env.example` (run `$ cp .env.example .env`)
  - If you are a core contributor you can install [Vercel CLI](https://vercel.com/docs/cli) and run `$ vercel link` and then `$ vercel env pull .env`

> - Open the `.env` file and update the values if needed based on the commented instructions

Reviewer 2: Maybe we move this to be after open localhost:300 and modify to 

```suggestion
- Optional: Open the  `.env` file and update excluded values if the functionality is needed for the feature your working on
```

Author: I agree we should add it, but I don't get why remove this one here, they have different purposes

Author: They do need to fill everything that is blank and not bellow the "Optional" title, I'll make it clearer with something like:
```Open the `.env` file and fill all the blank values that are not bellow the `OPTIONAL` title. Follow the commented instructions for each one```

Reviewer 2: Gotcha, didn't realize they were required 👍

- Start the local db by running `$ docker-compose -f dev.docker-compose.yml up -d`
  - We use docker-compose to run services both in local dev and in e2e testing

> - Run `$ npm run initial`, that includes all setup commands for first-time local development

Reviewer 2: ```suggestion
- Run `$ npm run initial`, this includes all setup commands for first-time local development
```

- In different terminal tabs, run `$ npm run dev` and `$ npm run inngest`
  - Both the development server AND Inngest are required for a fully functional website
- Open [http://localhost:3000](http://localhost:3000) with your browser to see the results
- Implement your changes, tests and documentations
- Commit your changes
- Submit your PR for review

### Pre-commit audits & testing

- Before committing changes, please audit your code locally:

  - `npm run lint` - audits code against ESLint and Prettier
  - `npm run typecheck` - audits code against TypeScript

>   - `npm run test` - execute Jest tests

Reviewer 2: ```suggestion
  - `npm run test` - executes Jest tests
```

  - `npm run audit` - executes all 3 commands above concurrently

>     - If `audit` fails, then it is very likely that your PR will not pass GitHub CI

Reviewer 2: ```suggestion
    - If `audit` fails, your PR will not pass GitHub CI
```

- Likewise, test E2E flows as necessary:
  - Make sure to stop your dev docker compose before running tests, we use a different instance of mysql for cypress and it will cause port conflicts
  - `npm run e2e:run` - starts up [Cypress](https://www.cypress.io/) UI
    - With the UI, you can run individual E2E tests under "Specs" and visually follow the flows
  - `npm run e2e:run-headless` - runs all our E2E tests consecutively via Cypress CLI

>     - Currently, this command only executes when new comments are made to main branch because of database dependencies - this may change in the future

Reviewer 2: we can get rid of this i think

    - Use tools like [DataGrip](https://www.jetbrains.com/datagrip/) or [DBeaver](https://dbeaver.io/) to analyze local MySQL data if necessary
  - Refer to `docs/Coding Conventions.md/Testing` for more information

### Useful development tips

- `npx prisma generate` - generates [Prisma ORM](https://www.prisma.io/) TypeScript definitions (_i.e._ DB client code) based on `prisma/prisma.schema`
  - Whenever you make Prisma schema changes, you should run the generate command
- `npx prisma studio` - spins up a full database UI for the database you are connected to
- `npm run db:seed` - resets the database you are connected to, then populates it with seed data
  - The database will be based on the `DATABASE_URL` provided in your local `.env` file - we will refer to this as "the database you are connected to"
- `npm run codegen` - generates TypeScript definitions for any GraphQL operations we have defined in our codebase
  - Whenever you make GraphQL query updates (_e.g._ our DTSI integration), you should run this command to update your TypeScript definitions

> - `npm run codegen:schemas` - if any of the GraphQL schemas changed for our 3rd-party API partners (_e.g._ https://www.dotheysupportit.com/), run this command to pull in the latest schemas

Reviewer 2: ```suggestion
- `npm run codegen:schemas` - if any of the GraphQL schemas change for our 3rd-party API partners (_e.g._ https://www.dotheysupportit.com/), run this command to pull in the latest updates
```

- `npm run intl:extract-compile` - generates initial translation files
- `npm run storybook` - view isolated [storybook stories](https://storybook.js.org/)

> - When checking out someone else's branch with changes - if you encounter TypeScript errors, run `npm run newbranch` to execute several common commands that may fix these errors

Reviewer 2: ```suggestion
-`npm run newbranch` - A useful command to run after checking out a newbranch or pulling the latest from `main` to ensure your local environment is configured correctly
```

### Updating the database schema

If you are not a core contributor, reach out to one to make any DB schema updates your PR needs. If you're a core contributor, follow this steps whenever you need to make Prisma Schema changes

- `npx prisma generate` - updates your Prisma TypeScript definitions
- Then, depending on your situation, run either of the following:
  - `npx prisma db push` - if your changes are _not_ breaking and you want to maintain the values in your database, then run this command; this will simply push the schema changes to the database instance you are connected to

>   - `npm run db:seed` - if you do have breaking changes, or if you just want to reset your database values, then run this command; this command force-pushes the schema updates to the database instance you are connected to, wiping all existing information and freshly populating the database with seed data

Reviewer 2: We should only include `npx prisma db push` as an option moving forward

> ### Running One Time Scripts On Testing or Production Environment

Reviewer 2: ```suggestion
### Running One Time Scripts in Testing or Production Environment
```

> On a testing or production environment, it is recommended to run the script through Inngest admin dashboard:

Reviewer 2: ```suggestion
In our testing or production environment, it is recommended to run a script through the Inngest admin dashboard:
```

- Select the function you want to run

> - Click the invoke button in the right corner

Reviewer 2: ```suggestion
- Click the invoke button in the top right corner
```

Author: They do need to fill everything that is blank and not bellow the "Optional" title, I'll make it clearer with something like:
```Open the `.env` file and fill all the blank values that are not bellow the `OPTIONAL` title. Follow the commented instructions for each one```
